libs/pilight/protocols/433.92/nexus.c

```diff
@@ -0,0 +1,257 @@
+	There is a 4000 µs long low state period before every 36 bits.
+
+	See also https://github.com/merbanan/rtl_433/blob/master/src/devices/nexus.c
+*/
+
+#include <math.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+
+#include "../../core/binary.h"
+#include "../../core/common.h"
+#include "../../core/dso.h"
+#include "../../core/gc.h"
+#include "../../core/log.h"
+#include "../../core/pilight.h"
+#include "../protocol.h"
+#include "nexus.h"
+
+#define PULSE_TOLERANCE 150
+#define MESSAGE_BITS 36
+#define RAW_LENGTH ((MESSAGE_BITS)*2) + 2
+
+enum PulseType {
+    START_P = 500,
+    ZERO_P = 1000,
+    ONE_P = 2000,
+    SYNC_P = 4000
+};
+
+typedef struct settings_t {
+    int id;
+    int channel;
+    int battery;
+    double temp;
+    double humi;
+    struct settings_t *next;
+} settings_t;
+
+static struct settings_t *settings = NULL;
+
+/**
+ * Validate whether a raw pulse length matches a known type
+ */
+static bool isValidPulse(uint16_t pulseLength, enum PulseType measureAgainst) {
+    return (pulseLength > (measureAgainst - PULSE_TOLERANCE)) && (pulseLength < (measureAgainst + PULSE_TOLERANCE));
+}
+
+static int validate(void) {
+    if (nexus->rawlen >= RAW_LENGTH && isValidPulse(nexus->raw[0], START_P) && isValidPulse(nexus->raw[RAW_LENGTH - 2], START_P)) {
+        // the first and second-to-last pulse should be a short one
+        return 0;
+    }
+    return -1;
+}
+
+static void parseCode(void) {
+    const int MAXBITS = nexus->rawlen / 2;
+    int id = 0, battery = 0, channel = 0;
+    double humi_offset = 0.0, temp_offset = 0.0;
+    double temperature = 0.0, humidity = 0.0;
+    int binary[MAXBITS];
+    int x = 0, i = 0;
+
+    // not sure why the check is needed, we use only the first 36 bits anyway
+    // if(nexus->rawlen>RAW_LENGTH) {
+    // 	logprintf(LOG_ERR, "nexus: parsecode - invalid parameter passed %d", nexus->rawlen);
+    // 	return;
+    // }
+
+    // decode pulses into bits
+    for (x = 1; x < nexus->rawlen - 1; x += 2) {
+        if (!isValidPulse(nexus->raw[x - 1], START_P)) {
+            return;
+        }
+        if (isValidPulse(nexus->raw[x], ONE_P)) {
+            binary[i++] = 1;
+        } else if (isValidPulse(nexus->raw[x], ZERO_P)) {
+            binary[i++] = 0;
+        } else {
+            return;
+        }
+    }
+
+    // bit 10 should be 0 and bits 25-28 should be 1
+    if (binary[9] != 0) {
+        return;
+    }
+    if (binary[24] != 1 || binary[25] != 1 || binary[26] != 1 || binary[27] != 1) {
+        return;
+    }
+
+    // decode bits into data
+    id = binToDecRev(binary, 0, 7);
+    battery = binary[8] ? 100 : 0;
+    channel = binToDecRev(binary, 10, 11);
+    temperature = (double)binToSignedRev(binary, 12, 23);
+    humidity = (double)binToDecRev(binary, 28, 35);
+
+    temperature /= 10;
+
+    struct settings_t *tmp = settings;
+    while (tmp) {
+        if (tmp->id == id) {
+            humi_offset = tmp->humi;
```

Comment:
You can just do the math for temperature and humidity here directly. No need to store them into temp. variables.

Reply:
I just copied what the others were doing. Is this offset something that can be set from the settings/UI? I have no idea what it does.

Reply:
I know, but we learn :smile:

I can be done more efficiently. These values can be set the config indeed.